native/src/widget/text_input.rs

```diff
@@ -347,19 +308,18 @@ where
                     && self.state.is_pasting.is_none()
                     && !c.is_control() =>
             {
-                let cursor = self.state.cursor_position(&self.value);
-
-                match cursor {
-                    Cursor::Index(index) => {
-                        self.value.insert(index, c);
-                    }
-                    Cursor::Selection { .. } => {
-                        self.state.move_cursor_left(&self.value);
-                        self.value.remove_many(cursor.left(), cursor.right());
-                        self.value.insert(cursor.left(), c);
-                    }
+                if !self.state.cursor.is_selection() {
+                    self.value.insert(self.state.cursor.end(), c);
+                } else {
+                    self.value.remove_many(
+                        self.state.cursor.left(),
+                        self.state.cursor.right(),
+                    );
+                    self.state.cursor.move_left();
+                    self.value.insert(self.state.cursor.end(), c);
                 }
-                self.state.move_cursor_right(&self.value);
+
+                self.state.cursor.move_right(&self.value);
```

Comment:
`Cursor` and `Value` are very coupled when we perform text edition. They always change at the same time.

I think it may be a good idea to play with the concept of an `Editor` (name subject to change). Maybe something like:

```rust
struct Editor<'a> {
    value: &'a mut Value,
    cursor: &'a mut Cursor,
}
```

And then implement all the text editing methods in this context. Then, whenever we need to edit a value, we will only be able to do it through an `Editor`, which will take care of the `Cursor` for us too.

Reply:
Hm, i guess that's a good idea huh. I'll look into it.

Reply:
What exactly should be the scope of the Editor?
Should it be used only where we edit a value (delete/insert), so that text_input still has a value field and a cursor in its state.
Or should it handle the text selection too, so that we could replace the cursor with the editor in the text_input state?
Or does it make sense to somehow move the Value completely into the editor (if this is even possible) so that we use the editor for everything related to the value, for example just getting the actual value of it (on top of editing/setting it)

Reply:
Alright, so I wasn't really able to get something done that either just works or actually makes sense. Because of that I'll close this Conversation thingy here for now. But I'm still open for suggestions or "bigger" hints in the right direction if you still want to have such thing implemented (maybe my problem here is the notation of the struct in your example idk). I mean we can still talk about it or you can ofc implement it yourself :P